### ISSUE TITLE
Multiple issues related to minfication, merging and bundling js

### ISSUE THREAD

**Issue by gewaechshaus** [Strategery-Inc/Magento2-InfiniteScroll#10]
Hey there,

I did a composer based installation in a Magento 2.1.4 system. However, oob there seems to be a problem (maybe related to combine/merge/minify js) with an undefined object

    window.iasConfig is undefined

See everything frontend related here:
https://www.makerdise.com/3d-druck/ersatzteile.html

I just noticed that there are other missing objects related to the window object.

Best
Jan

**Comment by gewaechshaus** [comment]
It seems like the spinnerSrc: base64 encoded image, which is a very long string, seems to get cutoff without the closing 

     ' 

during the merge/minify process.

https://github.com/Strategery-Inc/Magento2-InfiniteScroll/blob/master/view/frontend/templates/js.phtml#L36

I did set an image in the backend configuration to make the condition match before.

**Comment by gewaechshaus** [comment]
ReferenceError: IASPagingExtension is not defined

**Comment by gewaechshaus** [comment]
Definitly related to bundling... I am trying to exclude the .js from it, without success atm

   
     <exclude>
        <item type="file">Strategery_Infinitescroll::js/infinitescroll.js</item>
    </exclude>

**Comment by gsomoza** [comment]
Thanks for reporting this! @damian-pastorini will take a look ASAP. Meanwhile, if you find a fix, please feel free to submit a PR!

**Comment by gsomoza** [comment]
@damian-pastorini TBH I don't see why we have to use PHP to echo that? It seems like that whole string can go straight into the JS. And there might even be better options than doing that in the first place.

**Comment by gewaechshaus** [comment]
I enabled the debug flag on www.makedise.com... I tried all possible variations to exclude it from bundling, actually without success.

**Comment by gsomoza** [comment]
I think the problem is not with bundling, but with how we're trying to import and use some of the IAS extensions (it's currently failing to use the `IASPagerExtension`).

We should look into how to properly use IAS using RequireJS: we should distribute the original files instead of the minified versions and require them through RequireJS - since Magento 2 will take care of minifying them, etc, using their toolchain.

**Comment by gsomoza** [comment]
This may even be an opportunity to NOT distribute IAS using this extension, but rather let `npm` handle that (assuming that IAS is available on `npm`).

**Comment by gewaechshaus** [comment]
@gsomoza - if I disable minification, merging and bundling the module works as expected, but you're right.

https://www.npmjs.com/package/jquery-ias

**Comment by gewaechshaus** [comment]
I just uploaded 
https://github.com/webcreate/infinite-ajax-scroll/blob/master/src/jquery-ias.js 
to the related folder and changed 
https://github.com/Strategery-Inc/Magento2-InfiniteScroll/blob/master/view/frontend/requirejs-config.js#L18
to
       jqueryIas: 'Strategery_Infinitescroll/js/jquery-ias'

Did a setup:static-content:deploy ... and flushed/cleaned all caches, but that doesn't seem to affect the IAS...

**In fact, the current version is 16kb compared to 18kb of the older minified version that comes with the module.**

**Comment by gewaechshaus** [comment]
Just a note, no solution to fix the issue:
The < exclude > does work in a custom child theme by setting this code in etc/view.xml


     <view xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Config/etc/view.xsd">
       <exclude>
           <item type="file">Strategery_Infinitescroll::js/infinitescroll.js</item>
           <item type="file">Strategery_Infinitescroll::js/jquery-ias.js</item>
       </exclude>
   </view>

**Comment by gsomoza** [comment]
> In fact, the current version is 16kb compared to 18kb of the older minified version that comes with the module.

That's probably because we're not distributing the latest version of IAS - another reason to delegate that to Magento.

On your last comment I didn't quite understand: does that view.xml work or not?

**Comment by gewaechshaus** [comment]
I does work in general (the files are minified by magento and included seperatly like modernizr.js), but this doesn't seem to affect the issue in my case.

**Comment by gsomoza** [comment]
Got it, thanks. I think if we refactor the way we include IAS like I mentioned above *everything* should work as it should. Definitely the way to go IMO

**Comment by gewaechshaus** [comment]
This would be the best way. Thanks a lot...

**Comment by gewaechshaus** [comment]
This would be the best way. Thanks a lot...

**Comment by gewaechshaus** [comment]
@gsomoza - are there any news related to this issue?

**Comment by gsomoza** [comment]
Not yet, and it's possible it will take a while before we can get to it.. feel free to contribute a PR if you find the time!

**Comment by stijnbernards** [comment]
@gsomoza @gewaechshaus I've provided a fix for this issue, I hope this solves it for you :)

**Comment by gsomoza** [comment]
Nice, thanks! @damian-pastorini could you review that?

**Comment by gewaechshaus** [comment]
@gsomoza - will provide feedback asap! thanks a lot.

**Comment by gewaechshaus** [comment]
Hi @gsomoza, @stijnbernards 

many thanks, that does fix the issue indeed - take a look at
https://dev.makerdise.com/mechanik.html

I noticed another issue, maybe related to nginx conf and long urls
https://dev.makerdise.com/mechanik.html

Take a look at the console, there are 404 errors logged down... I am trying to trace this now.

Regards
Jan

**Comment by gsomoza** [comment]
Thanks @gewaechshaus !

**Comment by damian-pastorini** [comment]
Hi everyone, 
I just tested in a Magento 2.1.9 and merged some of the PR's, everything looks fine even after enabled/disabled the JS Merge, Bundling and Minify.
I didn't merged:
https://github.com/Strategery-Inc/Magento2-InfiniteScroll/pull/15/files#diff-9351d8171ad9c9d83cd82c302bb105ae 
Since that PR modify the original IAS library (which I would like to avoid), and even without this PR the extension worked fine.
Please give it a try to the last code we have in the master branch and let us know if the issues persist.

Thank you all for the provided fixes!

**Comment by gewaechshaus** [comment]
@damian-pastorini - I did drop the old installation and did a composer require strategery/magento2-infinitescroll:dev-master in a M2.2.1 installation... Still got the primary issue in this system as before, which is fixed here https://github.com/Strategery-Inc/Magento2-InfiniteScroll/pull/15/files#diff-9351d8171ad9c9d83cd82c302bb105ae
 
uncaught exception: ReferenceError: IASPagingExtension is not defined

Greetings
Jan

**Comment by damian-pastorini** [comment]
Hi @gewaechshaus , I'm not sure if the package for composer was updated, could you try with the code from here (GIT) please?
I don't have much time to check on this now, but I'll have it present for when I have some time.
Thanks!

**Comment by benjick** [comment]
Any workaround for this?
